### PR TITLE
Extract the variable stack to a separate file and test it

### PR DESCRIPTION
**Pull request by Mingun**
Also this change add additional safety checks to the code generator and improves errors if they was violated

### CODE REVIEW

File: lib/compiler/stack.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+  /**
+   * Assigns `exprCode` to the new variable in the stack, returns generated code.
+   * As the result, the size of a stack increases on 1.
+   *
+   * @param {string} exprCode Any expression code that must be assigned to the new variable in the stack
+   * @return {string} Assignment code
+   */
+  push(exprCode) {
+    const code = this.name(++this.sp) + " = " + exprCode + ";";
```
Comment (hildjj) [inline lib/compiler/stack.js:40]: Even though this is only used from generate-js at the moment, I was surprised to find JS code here.  a) it makes it harder to see all of the JS code in one place b) it makes it harder to use this class for other languages later.

If you decide to do something about this, then `this.type` may or may not still be needed in the class.

File: lib/compiler/stack.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
+"use strict";
+
+class Stack {
+  constructor(ruleName, varName, type) {
```
Comment (hildjj) [inline lib/compiler/stack.js:4]: constructor docs.  for example, it wasn't clear that "type" could be "let".

File: test/unit/compiler/stack.spec.js [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    it("throws error when stack pointer decreases", function() {
+      expect(() => stack.checkedLoop(0, () => stack.pop())).to.throw(Error,
+        "Rule 'rule', position 0: "
+        + "Body of a loop can't move the stack pointer "
+        + "(before: 0, after: -1)."
+      );
+    });
+  });
+});
```
Comment (hildjj) [inline test/unit/compiler/stack.spec.js:240]: Nice tests!

**Review by hildjj — COMMENTED** [review]
I think this is close; it just needs a decision about how js-specific stack.js should be.

File: lib/compiler/stack.js (hunk 1 above)
Comment (Mingun) [reply lib/compiler/stack.js:40]: Actually, this class can be used by almost any C-like code generator: C/C++/C#/JS/Java. I'm not consider it as a public API right now, although

File: lib/compiler/stack.js (hunk 2 above)
Comment (Mingun) [reply lib/compiler/stack.js:12]: Added

**Comment by Mingun** [comment]
I've added missing documentation for `Stack` and fix some spelling errors (and hopefully did not introduce new).

File: lib/compiler/stack.js (hunk 1 above)
Comment (hildjj) [reply lib/compiler/stack.js:48]: OK, that makes sense.  Later, if we want to be able to generate languages like Python, we can deal with `type` and semicolon being optional without breaking any code that relies on this interface.

**Pull request merged**